File: java/no_network_quiz/src/app/Quiz.java
package app;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.StandardOpenOption;
import java.nio.file.Files;
import java.util.Collections;

class Quiz {
    private static final Path ANSWERS_FILEPATH = Paths.get("answers.txt");
    private static final String TEACHER_NAME = "the teacher";
    private static final String OUTPUT_FORMAT = "%s: %s";

    private static final BufferedReader s_consoleReader = new BufferedReader(new InputStreamReader(System.in));

    public static void main(String[] a_args) throws IOException { 
        //while (true) {
        String name = askForName();
        Printouts.helloUser(name);

        if (!isTeacher(name)) {
            quizStudent(name);
        }
        else {
            printAnswerResults();
        }
        //}
    }

    private static String askForName() throws IOException {
        while (true) {
        	System.out.println("Hello, what is your name?");
        	final String name = s_consoleReader.readLine();
        	if (!name.contains(" ")) {
                System.err.println("Your full name, please");
                continue;
        	}
        	return name;
        }
    }


    private static int askForAnswerNumber(int a_minimumAllowed, int a_maximumAllowed) throws IOException {
        System.out.println("Which is the correct answer?");
        while (true) {
            String response = s_consoleReader.readLine();
            try {
                int value = Integer.parseInt(response);
                if ((value >= a_minimumAllowed) && (value <= a_maximumAllowed)) {
                    return value;
                }
            }
            catch (NumberFormatException ex) {
            }
            Printouts.reportAnswerNumberError(a_minimumAllowed, a_maximumAllowed);
        }
    }

    private static void quizStudent(String a_studentName) throws IOException {
        MultipleChoiceQuestions questionAndAnswers = new MultipleChoiceQuestions();

        String[] answers = questionAndAnswers.getShuffledAnswers();
        Printouts.questionAndAnswerOptions(questionAndAnswers.getQuestion(), answers);

        Printouts.emptyLine();

        int answerNumberStartingAt1 = askForAnswerNumber(1, answers.length);

        String rightOrWrong = (questionAndAnswers.isCorrectAnswer(answers[answerNumberStartingAt1 - 1]))
            ? "Right"
            : "Wrong";
        addAnswerToAnswersFile(a_studentName, rightOrWrong);
        System.out.println("I have recorded your response. Goodbye.");
    }

    private static boolean isTeacher(String _name) {
        return _name.equalsIgnoreCase(TEACHER_NAME);
    }

    private static void addAnswerToAnswersFile(String a_studentName, String a_resultDescription) throws IOException {
        String outputLine = String.format(OUTPUT_FORMAT, a_resultDescription, a_studentName);
        Files.write(ANSWERS_FILEPATH, Collections.singletonList(outputLine), new StandardOpenOption[]{ StandardOpenOption.CREATE, StandardOpenOption.APPEND });
    }

    private static void printAnswerResults() throws IOException {
        Printouts.printFile(ANSWERS_FILEPATH);
    }
}

File: csharp/no_network_quiz/app/Quiz.cs
using System;
using System.IO;
using System.Text;

class Quiz
{
    private static readonly string ANSWERS_FILEPATH = "answers.txt";
    private static readonly string TEACHER_NAME = "the teacher";
    private static readonly string OUTPUT_FORMAT = "{0}: {1}";

    public static void Main(string[] args)
    {
        //while (5 == 5)
        //{
            string name = AskForName();
            Printouts.HelloUser(name);

            if (!IsTeacher(name))
            {
                QuizStudent(name);
            }
            else
            {
                PrintAnswerResults();
            }
        //}
    }

    private static string AskForName()
    {
        bool isLegalValue = false;
        string name = "";

        while (!isLegalValue)
        {
            isLegalValue = true;
            Console.WriteLine("Hello, what is your name?");
            name = Console.ReadLine() ?? "";

            int nameLength = name.Length;
            if (nameLength == 0)
            {
                Console.Error.WriteLine("Name cannot be empty");
                isLegalValue = false;
            }
            else if (!name.Contains(" "))
            {
                Console.Error.WriteLine("You must give your FULL name.");
                isLegalValue = false;
            }
        }

        return name;
    }

    private static int AskForAnswerNumber(int minimumAllowed, int maximumAllowed)
    {
        Console.WriteLine("Which is the correct answer?");
        while (true)
        {
            string response = Console.ReadLine() ?? "";
            try
            {
                int value = Int32.Parse(response);
                if (value >= minimumAllowed && value <= maximumAllowed)
                {
                    return value;
                }
                Printouts.ReportAnswerNumberError(minimumAllowed, maximumAllowed);
            }
            catch (FormatException)
            {
                Printouts.ReportAnswerNumberError(minimumAllowed, maximumAllowed);
            }
        }
    }

    private static void QuizStudent(string studentName)
    {
        MultipleChoiceQuestions questionAndAnswers = new MultipleChoiceQuestions();

        string[] answers = questionAndAnswers.GetShuffledAnswers();
        Printouts.QuestionAndAnswerOptions(questionAndAnswers.Question, answers);

        Printouts.EmptyLine();

        int answerNumberStartingAt1 = AskForAnswerNumber(1, answers.Length);

        string rightOrWrong = questionAndAnswers.IsCorrectAnswer(answers[answerNumberStartingAt1 - 1])
            ? "Right"
            : "Wrong";
        AddAnswerToAnswersFile(studentName, rightOrWrong);

        Printouts.ThankYou();
    }

    private static bool IsTeacher(string name)
    {
        return name.Equals(TEACHER_NAME, StringComparison.OrdinalIgnoreCase);
    }

    private static void PrintAnswerResults()
    {
        if (File.Exists(ANSWERS_FILEPATH))
        {
            Printouts.PrintFile(ANSWERS_FILEPATH);
        }
        else
        {
            Console.WriteLine("No answer results found.");
        }
    }

    private static void AddAnswerToAnswersFile(string studentName, string rightOrWrong)
    {
        string content = String.Format(OUTPUT_FORMAT, rightOrWrong, studentName);
        File.AppendAllText(ANSWERS_FILEPATH, content + Environment.NewLine, Encoding.UTF8);
    }
}
